lib/cli/daemoncommand.cpp

```diff
@@ -432,6 +432,21 @@ static void WorkerSignalHandler(int num, siginfo_t *info, void*)
 		case SIGINT:
 		case SIGTERM:
 			if (info->si_pid == 0 || info->si_pid == l_UmbrellaPid) {
+				/*
+				 * Trigger application shutdown signal to mainly signify the api listener
+				 * not to accept any API requests anymore
+				 */
+				Application::OnShutdownRequested();
+
+				/*
+				 * Let the worker process flow till all outstanding "Process" tasks are finished to
+				 * prevent the Process worker threads from being destroyed prematurely e.g. before the
+				 * ongoing package stage updates are completed. Otherwise, we may end up with some
+				 * corrupted stages without startup logs and status file.
+				 */
+				while(Process::GetPendingTasks().load() != 0) {
+				}
```

Comment:
You can't block in a signal handler.

Reply:
Where would be the appropriate place to block? Inside the `Application::RunEventLoop()`?

Reply:
> Hanging plugins would block shutdown.

👇 
"Though it is up to the caller (which calls the `Process::Run()` member method) to decide which operation is critical and which not."

If you as a programmer decide to mark such plugin checks as important, then you obviously brought it on yourself!